leader-election-impl/src/main/java/com/palantir/leader/proxy/AwaitingLeadershipProxy.java

```
@@ -114,13 +113,8 @@ protected Object handleInvocation(Object proxy, Method method, Object[] args) th
                     // treat a repeated NO_QUORUM as NOT_LEADING; likely we've been cut off from the other nodes
                     // and should assume we're not the leader
                     if (leading == StillLeadingStatus.NOT_LEADING || leading == StillLeadingStatus.NO_QUORUM) {
-                        return Futures.submitAsync(
-                                () -> {
-                                    leadershipStateManager.invalidateStateOnLostLeadership(
-                                            leadershipToken, null /* cause */);
-                                    throw new AssertionError("should not reach here");
-                                },
-                                executionExecutor);
+                        return Futures.immediateFailedFuture(leadershipStateManager.invalidateStateOnLostLeadership(
```

I don't believe this is correct. Whilst the return type is the `NotCurrentLeaderException`, this method does not return the exception, but throws at evaluation. This means that instead of returning a future that fails, the handleInvocation method itself throws the exception, which is not what's intended here.

I wrote a mini test: 
```
    @Test
    public void areTheyTheSame() {
        Future<?> future = Futures.immediateFailedFuture(returnException());
        assertThat(future).isDone();
    }

    RuntimeException returnException() {
        return new RuntimeException("This is a test");
    }

    RuntimeException throwException() {
        throw new RuntimeException("This is a test");
    }
```

using returnException, the test passes. With throwException, the test fails as the method threw an exception, rather than the future failing

---

I'll admit I'm a little surprised / concerned that none of our tests caught this.

---

good catch, will update and look into test coverage